From 543e3df4090003be1f891d12f4db88843e95cfd5 Mon Sep 17 00:00:00 2001
From: Ikey Doherty <ikey@serpentos.com>
Date: Wed, 27 Nov 2024 02:40:29 +0000
Subject: [PATCH 03/10] boulder: Allow RPATH dependencies across package
 boundaries

We may also hit the case where a valid RPATH-bound dependency is
used at build-time, and then need that at runtime via automatic
dependencies.

This is seen in gnome-shell, which has RPATH-bound dependencies
on the mutter package.

Signed-off-by: Ikey Doherty <ikey@serpentos.com>
---
 boulder/src/package/analysis/handler/elf.rs | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/boulder/src/package/analysis/handler/elf.rs b/boulder/src/package/analysis/handler/elf.rs
index 7d9f935..c9350e5 100644
--- a/boulder/src/package/analysis/handler/elf.rs
+++ b/boulder/src/package/analysis/handler/elf.rs
@@ -164,8 +164,10 @@ fn parse_dynamic_section(
         for offset in needed_offsets {
             if let Ok(name) = strtab.get(offset) {
                 let rpath_name = rpaths.iter().find_map(|rpath| {
-                    let p = root_dir.to_owned() + "/" + rpath + "/" + name;
-                    let path = Path::new(&p);
+                    let local_p = root_dir.to_owned() + "/" + rpath + "/" + name;
+                    let native_p = rpath.to_owned() + "/" + name;
+                    let path = Path::new(&local_p);
+                    let native_path = Path::new(&native_p);
                     if path.exists() {
                         Some(
                             Path::new("/")
@@ -175,6 +177,8 @@ fn parse_dynamic_section(
                                 .skip(3)
                                 .collect::<PathBuf>(),
                         )
+                    } else if native_path.exists() {
+                        Some(Path::new(rpath).join(name).components().skip(3).collect::<PathBuf>())
                     } else {
                         None
                     }
-- 
2.47.1

